### ISSUE TITLE
[Question] Plotting and fitting issues with Product Kernel

### ISSUE THREAD

**Issue by LogosSama** [cornellius-gp/gpytorch#1767]
Hi , i'm currently working on a project. The aim of my project is to predict the Global Horizontal Irradiance of the sun,  using a dataset.
My dataset is CSV excel files which with contains the measures take each 10 min from April to July.
The dataset contains 2 columns
-	One for the time measure
-	The second corresponding to the  GHI
I’ve used to kernels: a periodic kernel and a RQ kernel.
I have to issues, which in my opinion are correlated:
- First : the plotting of my data seems incorrect. Look at the star which represent the dataset. For the same period of time, there is multiple measurement of GHI. This is impossible given that the measurement is done each 10 min.
- The second problem is that, my GP model doesn’t fit the values in the test phase.

The first graph is obtained at the end of the code, after the training and validation phase.
The second graph is just a zoom . And we can see the problem that i have with plotting

Any idea please ?

```


import math
import torch
import matplotlib.pyplot as plt
import pandas as pd
import gpytorch
import numpy as np
import datetime as dt

%matplotlib notebook
%matplotlib inline
%load_ext autoreload
%autoreload 2

# ########################################################
# Train data
X_train = pd.read_csv('xtraining_matlab.csv')
X_train = torch.FloatTensor(X_train.xtraining.to_numpy())

Y_train = pd.read_csv('ytraining_matlab.csv')
Y_train = torch.FloatTensor(Y_train.ytraining.to_numpy())

# ########################################################
Test data
X_test = pd.read_csv('xtest_matlab.csv')
X_test = torch.FloatTensor(X_test.xtest.to_numpy())

Y_test = pd.read_csv('ytest_matlab.csv')
Y_test = torch.FloatTensor(Y_test.ytest.to_numpy())


# ########################################################
#Plotting
plt.figure(figsize=(8,6))
plt.plot(X_train,Y_train)
plt.xlabel('Observation')
plt.ylabel('GHI')
plt.show()

# ########################################################
# ########################################################

# 
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self,X_train, Y_train,likelihood):
        super(ExactGPModel, self).__init__(X_train, Y_train, likelihood)
        self.mean_module = gpytorch.means.ZeroMean()
        
        PeriodicKernel =  gpytorch.kernels.ScaleKernel(gpytorch.kernels.PeriodicKernel())
        RQKernel =  gpytorch.kernels.ScaleKernel(gpytorch.kernels.RQKernel())
        
        self.covar_module = gpytorch.kernels.ProductKernel(PeriodicKernel,RQKernel)
        
       
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x,covar_x)
    


# initialize likelihood and model
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(X_train, Y_train, likelihood)



##################################################################
model.likelihood.initialize(noise=0.1)

# Periodic Kernel initialisation
model.covar_module.kernels[0].initialize(outputscale=0.4972)
model.covar_module.kernels[0].base_kernel.initialize(lengthscale=2.2583)
model.covar_module.kernels[0].base_kernel.initialize(period_length=0.9987)

##################################################################
# RQ Kernel Initialisation 
model.covar_module.kernels[1].initialize(outputscale=0.2159)
model.covar_module.kernels[1].base_kernel.initialize(lengthscale=16.3855)
model.covar_module.kernels[1].base_kernel.initialize(alpha=0.0431)

##################################################################
# Print initialised values
print('Noise :',model.likelihood.noise_covar.noise,"\n")
print('Periodic outputscale :',model.covar_module.kernels[0].outputscale)
print('Peridoc Kernel lengthscale:',model.covar_module.kernels[0].base_kernel.lengthscale)
print('périod_length :',model.covar_module.kernels[0].base_kernel.period_length,"\n")


print('RQ outputscale :',model.covar_module.kernels[1].outputscale)
print('Alpha : ',model.covar_module.kernels[1].base_kernel.alpha)
print('RQ lengthscale:',model.covar_module.kernels[1].base_kernel.lengthscale)



##################################################################
##################################################################
# Training model
model.train()
likelihood.train()

optimizer = torch.optim.Adam([
    {'params': model.parameters()},
], lr=0.55)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)


##################################################################
##################################################################
training_iter = 200

for i in range(training_iter):
    # Zero gradients from previous iteration
    optimizer.zero_grad()
    # Output from model
    output = model(X_train)
    
    
    # the way the PyTorch optimizer works is that it looks for a variable called loss,
    # which also has a gradient (calc by calling backward), and tries to reduce it.
    loss = -mll(output, Y_train)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f   L(Period): %.3f  périod_length: %.3f  L(RQ): %.3f  Alpha: %.3f  noise: %.3f' % (
        i + 1, training_iter, loss,
        model.covar_module.kernels[0].base_kernel.lengthscale,
        model.covar_module.kernels[0].base_kernel.period_length,
        model.covar_module.kernels[1].base_kernel.lengthscale,
        model.covar_module.kernels[1].base_kernel.alpha,
        model.likelihood.noise
    ))
    optimizer.step()

##################################################################
##################################################################

# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

f_preds = model(X_test)
f_mean = f_preds.mean

# Test points are regularly spaced along [0,1]
# Make predictions by feeding model through likelihood

with torch.no_grad(), gpytorch.settings.fast_pred_var():
    observed_pred = likelihood(model(X_test))
    observed_pred_train = likelihood(model(X_train))
##################################################################
##################################################################

 with torch.no_grad():
    # Initialize plot
    f, ax = plt.subplots(1, 1, figsize=(10, 7))
    
    # Get upper and lower confidence bounds
    lower, upper = observed_pred.confidence_region()
    
    # Plot training data as black stars
    ax.plot(X_train,Y_train, 'k*')
    
    # Plot test data as red stars
    ax.plot(X_test,Y_test, 'r*')
    
    # Plot the result of the GPR training as blue line
    ax.plot(X_train,observed_pred_train.mean.numpy(), 'b')
    
    # Plot predictive means as red line
    ax.plot(X_test, observed_pred.mean.numpy(), 'g')
    
    # Shade between the lower and upper confidence bounds
    ax.fill_between(X_test, lower.numpy(), upper.numpy(), alpha=0.5)
    ax.legend(['Observed Data', 'Pred_Data','Training_GPR' ,'Test_GPR','95 % Confidence'])
```
![github01](https://user-images.githubusercontent.com/75796600/134660711-2d45579d-5f7c-436e-a7da-c6d8c638f50c.PNG)
![github02](https://user-images.githubusercontent.com/75796600/134660714-30ebd5e5-42ca-4faa-bc73-2d18d2fb4427.PNG)

**Comment by wjmaddox** [comment]
Without access to your data, it's somewhat impossible to tell if you do truly have repeated observations (which would cause the plotting issues). I would check that first. 

Second, you are capturing the period of the data at test time but there's "mean reversion" so to speak, which isn't that uncommon -- probably your estimated hypers on the RQ kernel are a bit underfit.

**Comment by LogosSama** [comment]
> Without access to your data, it's somewhat impossible to tell if you do truly have repeated observations (which would cause the plotting issues). I would check that first.
> 
> Second, you are capturing the period of the data at test time but there's "mean reversion" so to speak, which isn't that uncommon -- probably your estimated hypers on the RQ kernel are a bit underfit. edit: try a lower learning rate and training longer.

Sorry but i didn't understood what you said about the "mean's reversion".
I do not did a mean reversion in my code. No matter what i'll try to diminish the learning rate.

About Your first advice, here is my data :
- The 1st column is the raw data as date in matlab. 
- The 2nd column is the date transform with the method datenum (from matlab) and exported afterwards as training data under the label "X_train"
- The other columns are the values i get after transformation as a tensor and use of the print method: 
`X_train = torch.FloatTensor(X_train.xtraining.to_numpy())`
`Y_train = torch.FloatTensor(Y_train.ytraining.to_numpy())`
`for i in range (0,21):
    print('{0:.20f}'.format(X_train[i]))`

![xtrain_giyhub](https://user-images.githubusercontent.com/75796600/134951372-393f29ad-3d2b-4fff-b538-aa7fa52d6684.PNG)

**Comment by LogosSama** [comment]
I missclicked and closed the issue....

**Comment by LogosSama** [comment]
> Just by taking a look at the data, it seems like the rounding step is causing you to have repeated inputs, which is why the train x points are on top of each other. You can try training the entire model in double precision by `model = model.double()` for example.
> 
> By mean reversion, I mean the tendency of a GP with a constant / zero mean to eventually return to the constant / zero as the predictions head off the support of the data (e.g. you extrapolate).

Thank you. I used the method `model.double()` and change my `FloatTensor` to `DoubleTensor` and got a better result.
I lower the learning rate to `lr =0.375` and increase the iteration to 450 :
![github matlab result01](https://user-images.githubusercontent.com/75796600/135262402-875071a0-5fa1-451a-8312-36f345dc37e6.PNG)
![github matlab result02](https://user-images.githubusercontent.com/75796600/135262434-222c9bc2-922d-40e8-acbc-1e94dd4046eb.PNG)

Thank you for your help.
To get a better result i think i will try now, to use a sliding windows to update the information given to the GPR model.
